Групповой проект filmorate, спринт 13

masicool opened the pull request:
На ревью

albidgy committed:
add director 0.1:
1. removeOne() in BaseDbStorage
2. Director's controller, service, storage, row-mapper and entity
3. field directors in Film-entity
4. findByDirectorId() for Film's controller, service, storage
5. new tables directors and film_directors in schema.sql

//TODO:
1. check new logic by request-samples from postman collection
2. refactoring
3. write tests for new functional

albidgy committed:
add director 0.2:
1. Получение/добавление и удаление режиссеров конкретному фильму
при создании/обновлении фильма (делал по аналогии с авторским решением
для жанров фильма, чтобы сохранить единообразие подходов).
2. В валидаторе дат для releaseDate (MinimumDate) пришлось отключить
проверку на нахождение даты не в будущем, т.к. один из PUT-запросов
в Postman-коллекции прокидывет кейс, согласно которому, фильмы с релиз-датой
в будущем должны заводиться в систему.

//TODO:
1. ревью и возможно - рефакторинг

masicool committed:
Merge pull request #7 from masicool/add-director

Add director

valeriyek committed:
Выполнено первое задание: добавлена функциональность для удаления фильмов и пользователей по ID

valeriyek committed:
Убрал ResponseEntity, добавил  @ResponseStatus

valeriyek committed:
Убрал ResponseEntity, добавил  @ResponseStatus

ofermad committed:
Добавить вывод общих с другом фильмов с сортировкой по популярности

masicool committed:
Merge pull request #9 from masicool/add-remove-endpoint

add-remove

ofermad committed:
Исправть запрос, переделать код

ofermad committed:
Исправть запрос, переделать код

ofermad committed:
Исправть запрос, переделать код

masicool committed:
Добавить реализацию reviews

masicool committed:
Merge branch 'refs/heads/develop' into add-reviews

masicool committed:
Поправить ошибки после обновления develop

masicool committed:
Merge pull request #13 from masicool/add-common-films

add-common-films

Lamchi01 committed:
Merge pull request #14 from masicool/add-reviews

Добавить реализацию "Отзывы"

Lamchi01 committed:
Добавление ленты событий

Lamchi01 committed:
Редактирование после ревью
Удаление бесполезного DTO
Небольшая доработка модели
Добавление enum

Lamchi01 committed:
Исправление код стайла

Lamchi01 committed:
Исправление код стайла №2

Lamchi01 committed:
Добавил класс EventService, для уменьшения повторения в коде

Lamchi01 committed:
Merge pull request #15 from masicool/add-feed

Добавление ленты событий

ReViZe96 committed:
add search 0.1:
добавлен метод поиска фильмов по:
- названию
- имени режиссёра
- названию и имени режиссера

masicool committed:
Merge pull request #17 from masicool/add-search

add search 0.1:

masicool committed:
Добавить диаграмму и исправить наименование в README.md

valeriyek committed:
Реализация вывода самых популярных фильмов по жанру и годам.

valeriyek committed:
Merge pull request #19 from masicool/add-most-populars

Реализация вывода самых популярных фильмов по жанру и годам.

masicool committed:
Добавить функционал "Рекомендации"

masicool committed:
Merge pull request #20 from masicool/add-recommendations

Добавить функционал "Рекомендации"

masicool committed:
Fix all bugs, 1083 tests is complete!

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,49 @@
+
+    public DirectorController(DirectorService directorService) {
+        this.directorService = directorService;
+    }
```
Не рассматривали аннотацию `@RequiredArgsConstructor` вместо явных конструкторов? Это по сути boilerplate код и по своей сути не несет никакой логической нагрузки

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,49 @@
+    @DeleteMapping("/{id}")
+    public void deleteById(@PathVariable long id) {
+        directorService.deleteById(id);
+    }
```
Покройте логами уровня info все эндпоинты конструктора. Это позволит более детально отслеживать, дебажить и выявлять где именно и что пошло не так

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,49 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public Director createDirector(@Valid @RequestBody Director director) {
+        directorService.create(director);
+        return director;
```
Тут лучше возвращать созданный объект из метода `create()` 
`return director;` будет возвращать объект с нулевым айди

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -59,8 +60,41 @@ public Film deleteLikeFilm(@PathVariable long id, @PathVariable long userId) {
+    }
+
+    @GetMapping("/director/{directorId}")
+    public List<Film> findByDirectorId(@PathVariable long directorId, @RequestParam(defaultValue = "year") String sortBy) {
```
Рекомендую сделать для сортировки енам. Всегда можно посмотреть возможные варианты сортировки и при появлении новых, просто добавить константу и условие или case

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -59,8 +60,41 @@ public Film deleteLikeFilm(@PathVariable long id, @PathVariable long userId) {
+    @GetMapping("/search")
+    public List<Film> findFilms(@RequestParam String query,
+                                @RequestParam String by) {
+        if (!by.equals("director") && !by.equals("title")
```
Тут тоже будет супер реализовать енам

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/EventService.java:
```diff
@@ -0,0 +1,33 @@
+@RequiredArgsConstructor
+@Service
+public class EventService {
+    private final EventDbStorage storage;
```
Внедряйте зависимости через интерфейсы, а не через конкретные реализации, это позволит в дальнейшем проще управлять внедрением, задавая приоритет

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -91,11 +105,38 @@ public Film deleteLike(long filmId, long userId) {
-        log.trace("Получен запрос на получение {} популярных фильмов", count);
-        return filmStorage.popularFilms(count);
+    public List<Film> getPopularFilms(int count, Long genreId, Integer year) {
+        log.trace("Получен запрос на получение {} популярных фильмов с фильтрацией по жанру и году", count);
```
Обычно выполнение каких-то определенных функций логируется уровнем info

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,89 @@
+        Review review = reviewStorage.findById(reviewId);
+        User user = userStorage.findById(userId);
+        reviewStorage.deleteDislikeReview(review, user);
+    }
```
Можно объединить в один метод лайк/дизлайк и в один метод удаление лайка/дизлайка

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
В storage классе слишком много логики, для этого есть сервисный слой, он нужен для описания основной бизнес логики. В сторадж классах стоит держать только то, что связано с работой с сущностью из БД

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/mappers/DirectorRowMapper.java:
Вы можете реализовать метод-маппер в самих стораджах, чтобы не плодить классы и зависимости, как пример:
```java
static User makeUser(ResultSet rs, int rowNum) throws SQLException {
		return new User(rs.getLong("user_id"), rs.getString("email"), rs.getString("login"), rs.getString("name"), toLocalDate(rs.getDate("birthday")));
}
```

Ivanov-s-tmn reviewed (commented):
Всем привет! Извините за долгое ревью. 

В целом все отлично, вы молодцы, что следуете общему стиля проекта и стараетесь от него не отступать, но местами есть небольшие различия (местами наличие java docs, использование явных конструкторов/аннотаций для внедрений зависимостей, неравномерное логирование и тд.). При работе над одним проектов старайтесь всегда придерживаться общего стиля, чтобы проект со временем не превратился в франкенштейна (смеси из разных подходов разных разработчиков).

Рассмотрите возможность изменения структуры БД относительно хранения режиссеров, я описал это в одном из своих комментариев и предложил альтернативный вариант.

Старайтесь оставлять логи в основных частях проекта, контроллеры - сервисы - стораджи (репозитории), это позволит лучше отслеживать работу и проще отслеживать на каком этапе возникает проблема (обычно для такого используется уровень info).

Если будут вопросы по моим комментариям, обращайтесь)

Всем отличного дня!

ReViZe96 committed:
final review 0.1:
1. info logs for DirectorController
2. set trace-level to info-level for all logs in project
3. FilmFilters enum for findFilms()
4. FilmSortOrder enum for findByDirectorId()
5. Refactoring: redundant logic from storage-level moved to services

masicool committed:
Merge pull request #23 from masicool/add-director

final review 0.1:

masicool committed:
Fix after review

masicool committed:
Delete EventService

masicool committed:
Remove logic to LikeService from LikeStorage

masicool committed:
Remove logic to LikeService from LikeStorage

masicool committed:
Merge remote-tracking branch 'origin/develop' into develop

masicool committed:
Reformat code

masicool committed:
Reformat code 2

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -3,98 +3,108 @@
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
     public Film createFilm(@Valid @RequestBody Film film) {
+        log.info("Получен запрос на создание фильма");
```
Тут можно указать в логе, что пришло на вход и в логах других методов контроллера, где есть параметры аналогично

Ivanov-s-tmn reviewed (commented):
Всем привет! Все отлично, проект зачтен, удачи дальше и хороших выходных!
